test/test-parse.cpp

```diff
@@ -539,6 +539,111 @@ TEST(test_escape_first_character_in_keyword,
 }
+
+TEST(test_no_overflow, parser_depth_limit_not_exceeded) {
+  {
+    for (const char8 opening_paren : {u8'(', u8'['}) {
+      string8 opening_parens(parser::stack_limit - 1, opening_paren);
+      padded_string code(opening_parens);
+      spy_visitor v;
+      parser p(&code, &v);
+      bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+      EXPECT_TRUE(ok);
+    }
+  }
+
+  {
+    string8 left_curly(parser::stack_limit, '{');
+    padded_string code(left_curly);
+    spy_visitor v;
+    parser p(&code, &v);
+    bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+    EXPECT_TRUE(ok);
+  }
+
+  {
+    for (const string8 &exp :
+         {u8"function f(){", u8"() => {", u8"if(){", u8"await", u8"do{",
+          u8"for(){", u8"try{", u8"while(){", u8"with({}){"}) {
+      string8 exps;
+      exps.reserve(exp.size() * (parser::stack_limit));
+      for (int i = 0; i < parser::stack_limit; i++) {
+        exps.append(exp);
+      }
+      padded_string code(exps);
+      spy_visitor v;
+      parser p(&code, &v);
+      bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+      EXPECT_TRUE(ok);
+    }
+  }
+
+  {
+    string8 object_decl(u8"{x:");
+    string8 paren_object_decls(u8"(");
+    paren_object_decls.reserve(object_decl.size() * (parser::stack_limit - 2));
+    for (int i = 0; i < parser::stack_limit - 2; i++) {
+      paren_object_decls.append(object_decl);
+    }
+    padded_string code(paren_object_decls);
+    spy_visitor v;
+    parser p(&code, &v);
+    bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+    EXPECT_TRUE(ok);
+  }
+}
+
+#if QLJS_HAVE_SETJMP
+TEST(test_overflow, parser_depth_limit_exceeded) {
+  {
+    for (const char8 opening_paren : {u8'(', u8'{', u8'['}) {
+      string8 opening_parens(parser::stack_limit + 1, opening_paren);
+      padded_string code(opening_parens);
+      spy_visitor v;
+      parser p(&code, &v);
+      bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+      EXPECT_FALSE(ok);
+      ElementsAre(
+          ::testing::VariantWith<error_depth_limit_exceeded>(::testing::_));
+    }
+  }
+
+  {
+    for (const string8 &exp :
+         {u8"function f(){", u8"() => {", u8"if(){", u8"await ", u8"do{",
+          u8"for(){", u8"try{", u8"while(){", u8"with({}){"}) {
+      string8 exps;
+      exps.reserve(exp.size() * (parser::stack_limit + 1));
+      for (int i = 0; i < parser::stack_limit + 1; i++) {
+        exps.append(exp);
+      }
+      padded_string code(exps);
+      spy_visitor v;
+      parser p(&code, &v);
+      bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+      EXPECT_FALSE(ok);
+      ElementsAre(
+          ::testing::VariantWith<error_depth_limit_exceeded>(::testing::_));
+    }
+  }
+
+  {
+    string8 object_decl(u8"{x:");
+    string8 paren_object_decl(u8"(");
+    paren_object_decl.reserve(object_decl.size() * (parser::stack_limit + 1));
```

Comment:
Nit: Reserved capacity is off by 1:
```suggestion
    paren_object_decl.reserve(paren_object_decl.size() + object_decl.size() * (parser::stack_limit + 1));
```